File: backend/dormammu/agent/command_builder.py
from __future__ import annotations

from pathlib import Path

from dormammu.agent.models import AgentRunRequest, CliCapabilities, CommandPlan


def _resolve_prompt_mode(request: AgentRunRequest, capabilities: CliCapabilities) -> str:
    if request.input_mode != "auto":
        return request.input_mode
    if request.prompt_flag or capabilities.prompt_file_flag:
        return "file"
    if capabilities.prompt_arg_flag:
        return "arg"
    return "stdin"


def build_command_plan(
    request: AgentRunRequest,
    capabilities: CliCapabilities,
    *,
    prompt_path: Path,
) -> CommandPlan:
    prompt_mode = _resolve_prompt_mode(request, capabilities)
    argv = [str(request.cli_path)]
    stdin_input: str | None = None

    if prompt_mode == "file":
        prompt_flag = request.prompt_flag or capabilities.prompt_file_flag
        if prompt_flag is None:
            raise ValueError("No prompt file flag is available for file mode.")
        argv.extend([prompt_flag, str(prompt_path)])
    elif prompt_mode == "arg":
        prompt_flag = request.prompt_flag or capabilities.prompt_arg_flag
        if prompt_flag is None:
            raise ValueError("No prompt argument flag is available for arg mode.")
        argv.extend([prompt_flag, request.prompt_text])
    elif prompt_mode == "stdin":
        stdin_input = request.prompt_text
    else:
        raise ValueError(f"Unsupported prompt mode: {prompt_mode}")

    argv.extend(request.extra_args)
    return CommandPlan(argv=argv, prompt_mode=prompt_mode, stdin_input=stdin_input)

File: runtime/src/agent/commandBuilder.ts
export type InputMode = "auto" | "file" | "arg" | "stdin" | "positional";

export type CliCapabilities = {
  helpFlag: string;
  promptFileFlag: string | null;
  promptArgFlag: string | null;
  workdirFlag: string | null;
  helpText: string;
  helpExitCode: number;
  commandPrefix?: readonly string[];
  promptPositional?: boolean;
  presetKey?: string | null;
  presetLabel?: string | null;
  presetSource?: string | null;
  autoApprove?: AutoApproveInfo;
};

export type AutoApproveCandidate = {
  value: string;
  risk: string;
  source: string;
  summary: string;
};

export type AutoApproveInfo = {
  supported: boolean;
  requiresConfirmation: boolean;
  candidates: readonly AutoApproveCandidate[];
  notes: readonly string[];
};

export type AgentRunRequest = {
  cliPath: string;
  promptText: string;
  repoRoot: string;
  workdir?: string | null;
  inputMode?: InputMode;
  promptFlag?: string | null;
  extraArgs?: readonly string[];
  runLabel?: string | null;
};

export type CommandPlan = {
  argv: string[];
  promptMode: Exclude<InputMode, "auto">;
  stdinInput: string | null;
};

function resolvePromptMode(
  request: AgentRunRequest,
  capabilities: CliCapabilities
): Exclude<InputMode, "auto"> {
  const requestedMode = request.inputMode ?? "auto";
  if (requestedMode !== "auto") {
    return requestedMode;
  }
  if (request.promptFlag || capabilities.promptFileFlag) {
    return "file";
  }
  if (capabilities.promptArgFlag) {
    return "arg";
  }
  if (capabilities.promptPositional) {
    return "positional";
  }
  return "stdin";
}

export function buildCommandPlan(
  request: AgentRunRequest,
  capabilities: CliCapabilities,
  promptPath: string
): CommandPlan {
  const promptMode = resolvePromptMode(request, capabilities);
  const argv = [
    request.cliPath,
    ...(capabilities.commandPrefix ?? [])
  ];
  let stdinInput: string | null = null;
  const trailingArgs: string[] = [];

  if (request.workdir && capabilities.workdirFlag) {
    argv.push(capabilities.workdirFlag, request.workdir);
  }

  if (promptMode === "file") {
    const promptFlag = request.promptFlag ?? capabilities.promptFileFlag;
    if (!promptFlag) {
      throw new Error("No prompt file flag is available for file mode.");
    }
    argv.push(promptFlag, promptPath);
  } else if (promptMode === "arg") {
    const promptFlag = request.promptFlag ?? capabilities.promptArgFlag;
    if (!promptFlag) {
      throw new Error("No prompt argument flag is available for arg mode.");
    }
    argv.push(promptFlag, request.promptText);
  } else if (promptMode === "stdin") {
    stdinInput = request.promptText;
  } else if (promptMode === "positional") {
    trailingArgs.push(request.promptText);
  } else {
    throw new Error(`Unsupported prompt mode: ${promptMode}`);
  }

  argv.push(...(request.extraArgs ?? []), ...trailingArgs);
  return { argv, promptMode, stdinInput };
}
